jena-core/src/main/java/org/apache/jena/mem/GraphMemUsingHashMap.java

```diff
@@ -0,0 +1,774 @@
+        if(sList != null) {
+            if(sList.contains(t)) {
+                return; /*triple already exists*/
+            }
+        } else {
+            sList = new ArrayList<>(INITIAL_SIZE_FOR_ARRAY_LISTS);
+            this.triplesBySubject.put(sIdx, sList);
+        }
+        sList.add(t);
+        {
+            var pIdx = t.getPredicate().getIndexingValue().hashCode();
+            var pList = this.triplesByPredicate
+                    .computeIfAbsent(pIdx,
+                            key -> new ArrayList<>(INITIAL_SIZE_FOR_ARRAY_LISTS));
+            pList.add(t);
+        }
+        {
+            var oIdx = t.getObject().getIndexingValue().hashCode();
+            var oList = this.triplesByObject
+                    .computeIfAbsent(oIdx,
+                            key -> new ArrayList<>(INITIAL_SIZE_FOR_ARRAY_LISTS));
+            oList.add(t);
+        }
+    }
+
+    /**
+     * Remove a triple from the triple store. The default implementation throws
+     * a DeleteDeniedException; subclasses must override if they want to be able
+     * to remove triples.
+     *
+     * @param t triple to delete
+     */
+    @SuppressWarnings("java:S1199")
+    @Override
+    public void performDelete(Triple t) {
+        var sIdx = t.getSubject().getIndexingValue().hashCode();
+        var sList = this.triplesBySubject.get(sIdx);
+        if(sList == null) {
+            return; /*triple does not exist*/
+        }
+        if(sList.remove(t)) {
+            if(sList.isEmpty()) {
+                this.triplesBySubject.remove(sIdx);
+            }
+            {
+                var pIdx = t.getPredicate().getIndexingValue().hashCode();
+                var pList = this.triplesByPredicate.get(pIdx);
+                pList.remove(t);
+                if (pList.isEmpty()) {
+                    this.triplesByPredicate.remove(pIdx);
+                }
+            }
+            {
+                var oIdx = t.getObject().getIndexingValue().hashCode();
+                var oList = this.triplesByObject.get(oIdx);
+                oList.remove(t);
+                if (oList.isEmpty()) {
+                    this.triplesByObject.remove(oIdx);
+                }
+            }
+        }
+    }
+
+    /**
+     * Remove all the statements from this graph.
+     */
+    @Override
+    public void clear() {
+        super.clear(); /* deletes all triples --> could be done better but later*/
+        this.triplesBySubject.clear();
+        this.triplesByPredicate.clear();
+        this.triplesByObject.clear();
+    }
+
+    /**
+     * Answer true if the graph contains any triple matching <code>t</code>.
+     * The default implementation uses <code>find</code> and checks to see
+     * if the iterator is non-empty.
+     *
+     * @param triple triple which may be contained
+     */
+    @SuppressWarnings("java:S3776")
+    @Override
+    protected boolean graphBaseContains(Triple triple) {
+        final Node sm = triple.getSubject();
```

Comment:
Seems like a lot of duplicated code here.  Why not just `return find(triple.getSubject(), triple.getPredicate(), triple.getObject()).findAny().isPresent();` ?

Reply:
This implementation should deliver the best possible performance. 
find() creates iterators which are not needed here.

We are developing an event sourcing in memory graph database. Every transaction is written to a  new delta graph (not the jena Delta but a variation). Delta graphs are chained until the next snapshot. 
In this scenario Graph.contains() and Graph.find() are havily used when performing SPAQRL queries on long chains of deltas to support temporal queries.

Reply:
That's similar to my experience. The effect is significantly more noticeable with Streams.